compiler-core/src/dependency.rs

```diff
@@ -137,19 +134,47 @@ but it is locked to {locked_version}, which is incompatible.",
     Ok(requirements)
 }
 
-pub trait PackageFetcher {
-    fn get_dependencies(&self, package: &str) -> Result<hexpm::Package, Box<dyn StdError>>;
+#[derive(Debug)]
+pub enum PackageFetchError {
+    ApiError(hexpm::ApiError),
+    FetchError(String),
+    IncompatibleLockedVersion(String),
+}
+impl std::fmt::Display for PackageFetchError {
```

Comment:
We avoid implementing traits in this codebase, especially `std::fmt::Display` as being explicit about what the format is at point of writing clearer and less error prone.

What was the motivation for adding this? I couldn't find where it is used.

Reply:
`PackageFetchError` implements `std::fmt::Display` to be able to implement `std::error::Error`. The reason for the addition of this error is that `DependencyResolver` needs a concrete associated error type for it's errors. Previously it just erred with `Box<dyn Error>`. Is there a better way for me to resolve this issue?

Reply:
I changed this to use `thiserror` instead. I am unsure of how I should format the errors, so i kept them unformatted.